packages/core/src/types.ts

```diff
@@ -1984,18 +1985,53 @@ type Matches<TypegenEnabledArg, TypegenDisabledArg> = {
   (stateValue: TypegenDisabledArg): any;
 };
 
-export type StateValueFrom<TMachine extends AnyStateMachine> =
-  StateFrom<TMachine>['matches'] extends Matches<
-    infer TypegenEnabledArg,
-    infer TypegenDisabledArg
-  >
-    ? TMachine['__TResolvedTypesMeta'] extends TypegenEnabled
-      ? TypegenEnabledArg
-      : TypegenDisabledArg
-    : never;
+export type StateValueFrom<
+  TMachine extends AnyStateMachine
+> = StateFrom<TMachine>['matches'] extends Matches<
+  infer TypegenEnabledArg,
+  infer TypegenDisabledArg
+>
+  ? TMachine['__TResolvedTypesMeta'] extends TypegenEnabled
+    ? TypegenEnabledArg
+    : TypegenDisabledArg
+  : never;
 
 export type PredictableActionArgumentsExec = (
   action: ActionObject<unknown, EventObject>,
   context: unknown,
   _event: SCXML.Event<EventObject>
 ) => void;
+
+type _TagsFrom<TResolvedTypesMeta> = TResolvedTypesMeta extends TypegenEnabled
+  ? Prop<Prop<TResolvedTypesMeta, 'resolved'>, 'tags'>
+  : string;
+
+export type TagsFrom<T> = ReturnTypeOrValue<T> extends infer R
```

Comment:
I would follow the structure of `StateValueFrom` here. Your version might be a little bit easier to use at times - as it handles more input types but it feels like grabbing this stuff from the machine type is the most common use case. At the very least - both of those types should accept the same inputs and for now, I would just downsize the scope of this new helper to the existing scope of `StateValueFrom`.

On top of that, I would infer this from the `hasTag` argument and not from the `TResolvedTypesMeta`.

Reply:
That makes perfect sense! I'll get that updated.